### ISSUE TITLE
spec: Ambiguity of type identity with non-generic types containing type parameter fields

### ISSUE THREAD

**Issue by Merovius** [golang/go#65152]
### What did you do?

[Playground Link](https://go.dev/play/p/phJTMuCj-Z0)

```go
package main

import "fmt"

func main() {
	fmt.Println(F[int]() == F[int](), F[int]() != G(), G() == G(), F[int]() != F[string]())
}

func F[T comparable]() any {
	type x struct{ _ [0]T }
	return x{}
}

func G() any {
	type x struct{ _ [0]int }
	return x{}
}
```

### What did you see happen?

`true true true true`

### What did you expect to see?

`true true true true`. The output is clearly intuitively correct and the most sensible behavior of this program. However, I'm having trouble justifying it from the spec.

For equality of interfaces, [the spec says](https://tip.golang.org/ref/spec#Comparison_operators):

> Interface types that are not type parameters are comparable. Two interface values are equal if they have [identical](https://tip.golang.org/ref/spec#Type_identity) dynamic types and equal dynamic values or if both have value `nil`.

The program is constructed such that the dynamic types all have exactly one dynamic value, hence we need to check if the dynamic types are identical:

> A [named type](https://tip.golang.org/ref/spec#Types) is always different from any other type. Otherwise, two types are identical if their [underlying](https://tip.golang.org/ref/spec#Types) type literals are structurally equivalent; that is, they have the same literal structure and corresponding components have identical types. In detail: […]

In this case, all types involved are named types. So, the question is whether those named types are "other types" or not. Determining whether that is the case is the contention. [The only relevant answer I can find says](https://tip.golang.org/ref/spec#Type_definitions):

> A type definition creates a new, distinct type with the same [underlying type](https://tip.golang.org/ref/spec#Underlying_types) and operations as the given type and binds an identifier, the type name, to it. […] The new type is called a defined type. It is [different](https://tip.golang.org/ref/spec#Type_identity) from any other type, including the type it is created from.

As well as later:

> If the type definition specifies [type parameters](https://tip.golang.org/ref/spec#Type_parameter_declarations), the type name denotes a generic type. Generic types must be [instantiated](https://tip.golang.org/ref/spec#Instantiations) when they are used.

Lastly, [going back to Type identity](https://tip.golang.org/ref/spec#Type_identity), we also note:

> Two [instantiated](https://tip.golang.org/ref/spec#Instantiations) types are identical if their defined types and all type arguments are identical.

That's all the relevant spec pieces I can really find.

Now my argument:

`F[int]() == F[int]()` and `G() == G()`, demonstrating that the created type is not tied to a call, but to the type definition AST node.

`G() != F[int]()` demonstrates that different type definition AST nodes lead to non-identical types. Strengthening the assertion that the actual AST node determines type identity.

Buf `F[int]() != F[string]()` shows that *the same* type definition can lead to *different* types. Even though 1. the type definition is not generic (it has no type parameters) and 2. it is a named type, so its identity *should* solely be determined by its type definition.

To be clear, again: This is obviously the only sensible and expected way for this code to behave, but it seems there is a hole in the spec for this behavior.

**Comment by Merovius** [comment]
cc @griesemer

**Comment by atdiar** [comment]
That's a very good question.

I even think, perhaps erroneously, that the case F[int]()! = G() is arguably true but could also be false with my current reading of the spec.

I'm not sure I've read that right but the scoping rules around type declaration seem to speak in term of visibility not equality.

As far as I remember, named type equality is still the equality of the name (identifier), type definition literal, and package path. But it might have changed.

https://go.dev/ref/spec#Declarations_and_scope
https://go.dev/ref/spec#Type_identity

Just a remark, I have no answer.

**Comment by zephyrtronium** [comment]
> In this case, all types involved are named types. So, the question is whether those named types are "other types" or not. Determining whether that is the case is the contention. ...

https://go.dev/ref/spec#Instantiations mentions, "instantiating a function produces a new non-generic function." So, every unique instantiation of `F` is a different function. That means a type declared within is different (in the identity sense) for each instantiation. This is true even if the type parameter isn't used in the type: https://go.dev/play/p/y8f4lMuhDUj (note the change to `F`). I recall there was some discussion on social media about this some months ago, but I'm struggling to find references now.

I believe this is the fallacy in your argument:

> `F[int]() == F[int]()` and `G() == G()`, demonstrating that the created type is not tied to a call, but to the type definition AST node.

It is not solely the AST, i.e. the syntactic location, which determines type identity. All type parameters which dominate the type definition participate as well.

**Comment by Merovius** [comment]
@zephyrtronium Ah, that actually makes sense, thank you. The behavior for non-used type parameters feels counterintuitive to me, but I guess it does follow from the spec and at least now I know :) I'll close this.

**Comment by seebs** [comment]
```
func F[T comparable]() any {
	return struct{ _ [0]T }{}
}

func G() any {
	return struct{ _ [0]int }{}
}
```

So it's the *named* types having the slightly surprising behavior -- if you use unnamed types, these two are now identical.

**Comment by atdiar** [comment]
Yup, after rereading, seems that type names are unique and scoped, that's why they are all different as per the spec. 

And the type definition includes the list of type arguments indeed but for x, there was none anyway so it doesn't even matter here.

**Comment by griesemer** [comment]
Reopening this (with title change) as a documentation issue. Similar questions have come up repeatedly and the spec could probably benefit from more explicit prose or related examples to be very clear.

**Comment by go101** [comment]
dup of https://github.com/golang/go/issues/58573

**Comment by go101** [comment]
@griesemer 

> Reopening this (with title change) as a documentation issue.

Will this be documented as an unspecified behavior?

**Comment by griesemer** [comment]
This is implicitly specified because "each distinct instantiation of a generic function produces a distinct non-generic function; and each of those functions has its own distinct local types, very much like any other non-generic function". Leaving this one open as a documentation issue and closing https://github.com/golang/go/issues/58573.
